### PR TITLE
Adding WASM support

### PR DESCRIPTION
**Pull request by neal-erickson**
Allowing WASM to work correctly from both the Circuit construction approach as well as allowing it to be passed in on the CLI for PECOS round-tripping.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -161,3 +161,6 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+
+# OS X
+.DS_Store
```
Comment (qartik) [inline .gitignore:166]: Instead of adding here, it's a good idea to configure global gitignore on your own machine, see https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer and https://github.com/github/gitignore/blob/main/Global/macOS.gitignore

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -7,6 +7,7 @@ repos:
       - id: check-toml
       - id: check-yaml
       - id: check-added-large-files
+        args: ['--maxkb=2000']
```
Comment (qartik) [inline .pre-commit-config.yaml:10]: That seems like a really large file. Is it possible to use a text format like WAT [like pecos does](https://github.com/PECOS-packages/PECOS/tree/development/tests/integration/wat)? See https://github.com/WebAssembly/wabt for `wasm2wat`

Additionally, it will help to be able to see the source of the WASM file instead of trusting a binary blob.

File: pytket/phir/phirgen.py [hunk 3]
```diff
@@ -27,6 +27,8 @@
 
 logger = logging.getLogger(__name__)
 
+StrAnyDict = dict[str, Any]
```
Comment (qartik) [inline pytket/phir/phirgen.py:30]: Doesn't seem to save that many keystrokes, perhaps best to call it something like `JsonType` to be clear of the intention if we are using an alias anyway.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (neal-erickson) [reply .pre-commit-config.yaml:10]: It's a bit of a bear to use WAT for two reasons: (1) pytket can't load WAT files (2) programmatically converting WAT to wasm files has been proving a bit of a bear without manually invoking wat2wasm which is CLI and would require a lot of awkward shell script usage. I've found that wasmtime can probably make me a byte array, which I would then have to save to a file on the file system as pytket requires a wasm file specifically. All of which is a lot of work to avoid a 1.6MB file I copied from the pytket repository: https://github.com/CQCL/tket/blob/4fae6b93d6f412163e783151a91a1270a4567969/pytket/tests/classical_test.py#L399

File: .gitignore (hunk 1 above)
Comment (neal-erickson) [reply .gitignore:166]: Fair, although a "too good" gitignore is not much of a problem.

File: pytket/phir/phirgen.py (hunk 3 above)
Comment (neal-erickson) [reply pytket/phir/phirgen.py:30]: That's a better name. More than keystrokes, I wanted to signify the commonality and expectation of unified types across the file.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (qartik) [reply .pre-commit-config.yaml:10]: Fair.

> a lot of work to avoid a 1.6MB file

To be clear it's not just to avoid a big file, but also for readability and version tracking. I always tend to prefer adding a source file instead of a binary when possible.

Comment (neal-erickson) [reply .pre-commit-config.yaml:10]: Yeah for sure, I would much prefer wat, but I think that would blow out the scope of this story a bit.

File: tests/data/qasm/bell.qasm
Comment (qartik) [inline tests/data/qasm/bell.qasm:1]: Is this file used anywhere?

File: tests/test_api.py [hunk 4]
```diff
@@ -85,3 +87,23 @@ def test_qasm_to_phir(self) -> None:
...
+        c.add_wasm_to_reg("multi", w, [c0, c1], [c2])
+        c.add_wasm_to_reg("add_one", w, [c2], [c2])
+        c.add_wasm_to_reg("no_return", w, [c2], [])
+        c.add_wasm_to_reg("no_parameters", w, [], [c2])
+
+        c.add_wasm_to_reg("add_one", w, [c0], [c0], condition=c1[0])
+
+        phir = pytket_to_phir(c, QtmMachine.H1_1)
+
+        assert phir
```
Comment (qartik) [inline tests/test_api.py:109]: Could we add fine-grained tests here? Specifically, it will be nice to ensure the issues we identified in the chat last month about args and returns don't occur.

File: pytket/phir/phirgen.py [hunk 5]
```diff
@@ -27,6 +27,8 @@
 
 logger = logging.getLogger(__name__)
 
+JsonDict = dict[str, Any]
+
 PHIR_HEADER: dict[str, Any] = {
```
Comment (qartik) [inline pytket/phir/phirgen.py:32]:
```suggestion
PHIR_HEADER: JsonDict = {
```

Another similar change is at line 364. You could also consider updating `phirgen_parallel.py` while you are at it.

File: pytket/phir/cli.py [hunk 6]
```diff
@@ -57,19 +61,24 @@ def main() -> None:
     args = parser.parse_args()
 
     for file in args.qasm_files:
-        print(f"Processing {file}")  # noqa: T201
-        c = circuit_from_qasm(file)
-        tket_opt_level = int(args.tk)
-        rc = rebase_to_qtm_machine(c, args.machine, tket_opt_level)
```
Comment (qartik) [inline pytket/phir/cli.py:63]: Where are we doing this rebasing now?

File: ruff.toml [hunk 7]
```diff
@@ -81,6 +81,9 @@ ignore = [
     "PLR2004",  # Magic constants
     "PLR6301",  # Method * could be a function, class method, or static method
   ]
+"pytket/phir/cli.py" = [
```
Comment (qartik) [inline ruff.toml:84]: For single file ignores, I tend to prefer adding the ignore at the top of the file, like so:
```
# ruff: noqa: T201
```

File: pytket/phir/cli.py [hunk 8]
```diff
@@ -8,21 +8,18 @@
 
 # mypy: disable-error-code="misc"
```
Comment (qartik) [inline pytket/phir/cli.py:9]:
```suggestion
# mypy: disable-error-code="misc"
# ruff: noqa: T201
```
See comment in `ruff.toml`

**Review by qartik — CHANGES_REQUESTED** [review]

File: tests/test_api.py (hunk 4 above)
Comment (neal-erickson) [reply tests/test_api.py:109]: Can you remind me of what issues those were?

File: pytket/phir/cli.py (hunk 6 above)
Comment (neal-erickson) [reply pytket/phir/cli.py:63]: The gate set rebase happens in `pytket_to_phir`

**Review by qartik — COMMENTED** [review]
Few more small suggestions, I still need to test the PR locally on my machine.

**Comment by qciaran** [comment]
I primarily use `pytket.phir.api.pytket_to_phir()` or `pytket.phir.api.qasm_to_phir()`. It seems like these do not include wasm input. I see the CLI has the option to read in a wasm file. Also, when running wasm, I often directly supply the wasm byte object, so if possible it would be nice to either supply the byte object or file path.

**Comment by qciaran** [comment]
I get what I expect when running a simulation after generating PHIR from using `circuit_from_qasm_wasm` and `pytket_to_phir`.

File: tests/data/qasm/bell.qasm
Comment (qartik) [inline tests/data/qasm/bell.qasm:1]: Is this file needed?

File: tests/test_api.py [hunk 9]
```diff
@@ -24,8 +29,8 @@ class TestApi:
     def test_pytket_to_phir_no_machine(self) -> None:
         """Test case when no machine is present."""
         circuit = get_qasm_as_circuit(QasmFile.baby)
-
-        assert pytket_to_phir(circuit)
+        phir = pytket_to_phir(circuit)
+        PHIRModel.model_validate_json(phir)
```
Comment (qartik) [inline tests/test_api.py:33]: I believe `phirgen` is already doing model validation, so to improve these tests the most we can do is assert for specific JSON entries in them. But they don't need to be done in this PR.

File: tests/test_api.py [hunk 10]
```diff
@@ -84,4 +91,118 @@ def test_qasm_to_phir(self) -> None:
         measure q[1]->cr[0];
         """
 
-        assert qasm_to_phir(qasm, QtmMachine.H1_1)
+        phir = qasm_to_phir(qasm, QtmMachine.H1_1)
+        PHIRModel.model_validate_json(phir)
```
Comment (qartik) [inline tests/test_api.py:95]: Same as above in several tests.

File: tests/test_api.py [hunk 11]
```diff
@@ -84,4 +91,118 @@ def test_qasm_to_phir(self) -> None:
...
+        wasm_bytes = get_wat_as_wasm_bytes(WatFile.add)
+
+        phir_str = qasm_to_phir(qasm, QtmMachine.H1_1, wasm_bytes=wasm_bytes)
+        phir = json.loads(phir_str)
+        assert phir is not None
+
+        expected_metadata = {
+            "ff_object": (
+                "WASM module uid: 28c0194b91f1e24d6fc40ec480c026a5874661184"
+                "bcd411a61bd5d5383df5180"
```
Comment (qartik) [inline tests/test_api.py:125]: Can we document somewhere how these module uid's are generated? They seem magic to me.

**Comment by qartik** [comment]
Looking good, thanks for finding a solution for consuming wat files. It will be nice if we can consume either wasm or wat files through the CLI as well. Could you also update the README after making changes to the CLI?

File: tests/test_api.py (hunk 11 above)
Comment (neal-erickson) [reply tests/test_api.py:125]: I actually don't know how other than to look after the fact in the PHIR.

File: tests/test_api.py (hunk 9 above)
Comment (neal-erickson) [reply tests/test_api.py:33]: Ah, thought it was only getting used sometimes, I see that it always does it.

File: pytket/phir/api.py [hunk 12]
```diff
@@ -91,6 +93,19 @@ def qasm_to_phir(
     :param circuit: Circuit object to be converted
     :param qtm_machine: (Optional) Quantinuum machine architecture to rebase against
     :param tket_optimization_level: (Default=0) TKET circuit optimization level
+    :param wasm_bytes (Optional) WASM as bytes to include as part of circuit
     """
-    circuit = circuit_from_qasm_str(qasm)
+    circuit: Circuit
+    if wasm_bytes:
```
Comment (qciaran) [inline pytket/phir/api.py:99]: On Windows I am getting "PermissionError: [Errno 13] Permission denied" for the wasm temporary file.

Comment (qciaran) [reply pytket/phir/api.py:99]: I have had issues in the past with creating temporary files where the file was getting deleted after I wrote to it without the right settings. I don't know if that is what is going on here.

Comment (qciaran) [reply pytket/phir/api.py:99]: If it is any help, this is what worked for me in the past:
```python
with tempfile.NamedTemporaryFile(mode="w+",suffix=".qasm", delete=False) as temp_file:
    temp_file.write(program)
    temp_file.flush()
    temp_file_path = Path(temp_file.name)
    program = qasm2pytket_handler(temp_file_path, machine)
temp_file_path.unlink()
```

File: tests/test_api.py (hunk 11 above)
Comment (qartik) [reply tests/test_api.py:119]: I think I'd like to find out. Or perhaps we should not assert for module id if it's irrelevant.

**Review by qartik — APPROVED** [review]
LGTM.

**Review by qartik — CHANGES_REQUESTED** [review]
As noticed by @Asa-Kosto-QTM in #96, the tket opt level arg needs to be handled correctly here. Currently, showing up as a merge conflict.

**Comment by neal-erickson** [comment]
@qartik tket_opt_level fix was already in place due to CLI changes. `main` merged into this branch.

File: tests/test_api.py [hunk 13]
```diff
@@ -85,3 +91,111 @@ def test_qasm_to_phir(self) -> None:
...
+        measure q[0]->cr[0];
+        measure q[1]->cr[1];
+
+        cs = cr;
+        co = add(cr, cs);
+        """
+
+        wasm_bytes = get_wat_as_wasm_bytes(WatFile.add)
+
+        wasm_uid = hashlib.sha256(base64.b64encode(wasm_bytes)).hexdigest()
```
Comment (qartik) [inline tests/test_api.py:116]: @neal-erickson I referenced https://github.com/CQCL/tket/blob/6906704b0ad64d077dff5c71454652926f91ee12/pytket/pytket/wasm/wasm.py#L77-L81 to generate this `wasm_id`

File: tests/test_api.py [hunk 14]
```diff
@@ -85,3 +91,111 @@ def test_qasm_to_phir(self) -> None:
...
+
+            c.add_wasm_to_reg("add_one", w, [c0], [c0], condition=c1[0])
+
+            phir_str = pytket_to_phir(c, QtmMachine.H1_1)
+        finally:
+            Path.unlink(Path(wasm_file.name))
+
+        phir = json.loads(phir_str)
+
+        expected_metadata = {"ff_object": (f"WASM module uid: {w!s}")}
```
Comment (qartik) [inline tests/test_api.py:160]: Here, since we are in pytket land, we can just use the string representation of `WasmFileHandler` instance.

**Review by qartik — APPROVED** [review]
LGTM

**Comment by qartik** [comment]
> It will be nice if we can consume either wasm or wat files through the CLI as well.

I was trying to see if I can add the functionality to accept .wat files but noticed that currently even .wasm files fail with an error:

```sh
❯ cat wasm.qasm
OPENQASM 2.0;
include "qelib1.inc";

qreg q[2];
h q;
ZZ q[1],q[0];
creg cr[3];
creg cs[3];
creg co[3];
measure q[0]->cr[0];
measure q[1]->cr[1];

cs = cr;
co = add(cr, cs);

❯ phirc -v -w add.wasm wasm.qasm
Processing wasm.qasm
Including WASM from file add.wasm

PHIR to be simulated:
{"format": "PHIR/JSON", "version": "0.1.0", "metadata": {"source": "pytket-phir v0.2.2.dev35", "strict_parallelism": true}, "ops": [{"data": "qvar_define", "data_type": "qubits", "variable": "q", "size": 2}, {"data": "cvar_define", "data_type": "u32", "variable": "cr", "size": 3}, {"data": "cvar_define", "data_type": "u32", "variable": "cs", "size": 3}, {"data": "cvar_define", "data_type": "u32", "variable": "co", "size": 3}, {"//": "Parallel Rz(1)"}, {"qop": "RZ", "angles": [[1.0], "pi"], "args": [["q", 0], ["q", 1]]}, {"//": "Parallel PhasedX(0.5, 0.5)"}, {"qop": "R1XY", "angles": [[0.5, 0.5], "pi"], "args": [["q", 0], ["q", 1]]}, {"//": "Parallel RZ"}, {"block": "qparallel", "ops": [{"qop": "RZ", "angles": [[0.5], "pi"], "args": [["q", 0]]}, {"qop": "RZ", "angles": [[1.5], "pi"], "args": [["q", 1]]}]}, {"//": "ZZPhase(0.5) q[1], q[0];"}, {"qop": "RZZ", "angles": [[0.5], "pi"], "args": [[["q", 1], ["q", 0]]]}, {"mop": "Transport", "duration": [0.0, "ms"]}, {"//": "Parallel RZ"}, {"block": "qparallel", "ops": [{"qop": "RZ", "angles": [[1.5], "pi"], "args": [["q", 0]]}, {"qop": "RZ", "angles": [[0.5], "pi"], "args": [["q", 1]]}]}, {"qop": "Measure", "args": [["q", 0], ["q", 1]], "returns": [["cr", 0], ["cr", 1]]}, {"mop": "Transport", "duration": [0.0, "ms"]}, {"//": "CopyBits cr[0], cr[1], cr[2], cs[0], cs[1], cs[2];"}, {"cop": "=", "returns": [["cs", 0], ["cs", 1], ["cs", 2]], "args": [["cr", 0], ["cr", 1], ["cr", 2]]}, {"mop": "Transport", "duration": [0.0, "ms"]}, {"//": "WASM function=add args=['cr', 'cs'] returns=['co']"}, {"cop": "ffcall", "function": "add", "args": ["cr", "cs"], "metadata": {"ff_object": "WASM module uid: 48b455b2aba63187d5f1720a71762c3f5cfbd359c309c1fa91e5d7027b2e111a"}, "returns": ["co"]}, {"mop": "Transport", "duration": [0.0, "ms"]}]}

PECOS results:
Traceback (most recent call last):
  File "/Users/kartik/GH/CQCL/pytket-phir/.venv/bin/phirc", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pytket/phir/cli.py", line 84, in main
    print(HybridEngine(qsim="state-vector").run(program=phir, shots=10))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pecos/engines/hybrid_engine.py", line 160, in run
    self.initialize_sim_components(program, foreign_object)
  File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pecos/engines/hybrid_engine.py", line 101, in initialize_sim_components
    num_qubits = self.cinterp.init(program, foreign_object)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pecos/classical_interpreters/phir_classical_interpreter.py", line 95, in init
    self.check_ffc(self.program.foreign_func_calls, self.foreign_obj)
  File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pecos/classical_interpreters/phir_classical_interpreter.py", line 103, in check_ffc
    func_names = set(fobj.get_funcs())
                     ^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'get_funcs'
```

Do we know what may be going on? cc: @qciaran

**Comment by qciaran** [comment]
> > It will be nice if we can consume either wasm or wat files through the CLI as well.
> 
> I was trying to see if I can add the functionality to accept .wat files but noticed that currently even .wasm files fail with an error:
> 
> ```shell
> ❯ cat wasm.qasm
> OPENQASM 2.0;
> include "qelib1.inc";
> 
> qreg q[2];
> h q;
> ZZ q[1],q[0];
> creg cr[3];
> creg cs[3];
> creg co[3];
> measure q[0]->cr[0];
> measure q[1]->cr[1];
> 
> cs = cr;
> co = add(cr, cs);
> 
> ❯ phirc -v -w add.wasm wasm.qasm
> Processing wasm.qasm
> Including WASM from file add.wasm
> 
> PHIR to be simulated:
> {"format": "PHIR/JSON", "version": "0.1.0", "metadata": {"source": "pytket-phir v0.2.2.dev35", "strict_parallelism": true}, "ops": [{"data": "qvar_define", "data_type": "qubits", "variable": "q", "size": 2}, {"data": "cvar_define", "data_type": "u32", "variable": "cr", "size": 3}, {"data": "cvar_define", "data_type": "u32", "variable": "cs", "size": 3}, {"data": "cvar_define", "data_type": "u32", "variable": "co", "size": 3}, {"//": "Parallel Rz(1)"}, {"qop": "RZ", "angles": [[1.0], "pi"], "args": [["q", 0], ["q", 1]]}, {"//": "Parallel PhasedX(0.5, 0.5)"}, {"qop": "R1XY", "angles": [[0.5, 0.5], "pi"], "args": [["q", 0], ["q", 1]]}, {"//": "Parallel RZ"}, {"block": "qparallel", "ops": [{"qop": "RZ", "angles": [[0.5], "pi"], "args": [["q", 0]]}, {"qop": "RZ", "angles": [[1.5], "pi"], "args": [["q", 1]]}]}, {"//": "ZZPhase(0.5) q[1], q[0];"}, {"qop": "RZZ", "angles": [[0.5], "pi"], "args": [[["q", 1], ["q", 0]]]}, {"mop": "Transport", "duration": [0.0, "ms"]}, {"//": "Parallel RZ"}, {"block": "qparallel", "ops": [{"qop": "RZ", "angles": [[1.5], "pi"], "args": [["q", 0]]}, {"qop": "RZ", "angles": [[0.5], "pi"], "args": [["q", 1]]}]}, {"qop": "Measure", "args": [["q", 0], ["q", 1]], "returns": [["cr", 0], ["cr", 1]]}, {"mop": "Transport", "duration": [0.0, "ms"]}, {"//": "CopyBits cr[0], cr[1], cr[2], cs[0], cs[1], cs[2];"}, {"cop": "=", "returns": [["cs", 0], ["cs", 1], ["cs", 2]], "args": [["cr", 0], ["cr", 1], ["cr", 2]]}, {"mop": "Transport", "duration": [0.0, "ms"]}, {"//": "WASM function=add args=['cr', 'cs'] returns=['co']"}, {"cop": "ffcall", "function": "add", "args": ["cr", "cs"], "metadata": {"ff_object": "WASM module uid: 48b455b2aba63187d5f1720a71762c3f5cfbd359c309c1fa91e5d7027b2e111a"}, "returns": ["co"]}, {"mop": "Transport", "duration": [0.0, "ms"]}]}
> 
> PECOS results:
> Traceback (most recent call last):
>   File "/Users/kartik/GH/CQCL/pytket-phir/.venv/bin/phirc", line 8, in <module>
>     sys.exit(main())
>              ^^^^^^
>   File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pytket/phir/cli.py", line 84, in main
>     print(HybridEngine(qsim="state-vector").run(program=phir, shots=10))
>           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pecos/engines/hybrid_engine.py", line 160, in run
>     self.initialize_sim_components(program, foreign_object)
>   File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pecos/engines/hybrid_engine.py", line 101, in initialize_sim_components
>     num_qubits = self.cinterp.init(program, foreign_object)
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pecos/classical_interpreters/phir_classical_interpreter.py", line 95, in init
>     self.check_ffc(self.program.foreign_func_calls, self.foreign_obj)
>   File "/Users/kartik/GH/CQCL/pytket-phir/.venv/lib/python3.12/site-packages/pecos/classical_interpreters/phir_classical_interpreter.py", line 103, in check_ffc
>     func_names = set(fobj.get_funcs())
>                      ^^^^^^^^^^^^^^
> AttributeError: 'NoneType' object has no attribute 'get_funcs'
> ```
> 
> Do we know what may be going on? cc: @qciaran

My guess you are trying to run PHIR that is calling Wasm but not supplying the Wasm to PECOS via a "ForeignObj". It would be something like:
```python
from pecos.foreign_objects.wasmtime import WasmtimeObj
wasm_obj = WasmtimeObj("./wasm/add.wasm")
HybridEngine(...).run(phir, foreign_object=wasm_obj, ...) 
```

It might be good for me to make the `run()` interface a little nice by supplying a `wasm_bytes` option and it do all the work of preparing the object for you.

File: pytket/phir/cli.py [hunk 15]
```diff
@@ -7,22 +7,21 @@
 ##############################################################################
 
 # mypy: disable-error-code="misc"
+# ruff: noqa: T201
 
 from argparse import ArgumentParser
 from importlib.metadata import version
 
 from pecos.engines.hybrid_engine import HybridEngine  # type:ignore [import-not-found]
+from pecos.foreign_objects.wasmtime import WasmtimeObj  # type:ignore [import-not-found]
```
Comment (qartik) [inline pytket/phir/cli.py:16]: Just sent a PR to pecos at https://github.com/PECOS-packages/PECOS/pull/37 to avoid these type ignores.

**Review by qartik — APPROVED** [review]
looking good!

**Review by qciaran — APPROVED** [review]
Working for me/LGTM
